### PR TITLE
Add Flash Attention and FP16 Activation

### PR DESCRIPTION
**Pull request by Jungwon-Lee**
## Summary

This PR adds two complementary features to the CausalLM inference path:

1. **GEMM-based flash attention** — online-softmax attention for long-context prefill, with GQA support and causal/sliding-window block-skip
2. **FP16 activation end-to-end** — Q4_0 and Q6_K weight × FP16 activation GEMM/GEMV kernels (NEON), plus the layer-level fixes needed to run the complete forward pass with FP16 activations

---

### Algorithm

Implements flash attention (Phase 2 tiling, online softmax) for the MHA forward pass in `mha_core`. The implementation:

- Activates for prefill only (`step_size >= 32`); single-token decode keeps the existing per-row dot path
- Supports GQA: Q-head / KV-head groups are mapped correctly without materialising extra KV copies
- Applies causal masking and sliding-window masking with block-skip to avoid computing fully-masked (Q, K) tiles
- Config: `use_flash_attention` (default `true`) in `nntr_config.json`, propagated through `Transformer` / `Qwen3Transformer` → `mha_core`

### Kernels

The QK and AV GEMM primitives are promoted to `cpu_backend` library functions:

| Kernel | ISA | Direction | File |
|--------|-----|-----------|------|
| `hgemm_f16xf16_f32_fmlal` | ARMv8.2-A FMLAL | FP16 × FP16 → FP32 (QK) | `cpu_backend/arm/hgemm/` |
| `hsgemm_fp16bits_avx2` | AVX2 + F16C + FMA | FP32 × FP16-bits → FP32 (QK, AV) | `cpu_backend/x86/avx2_impl.cpp` |

On x86, K and V are read as raw `uint16_t` FP16-bits directly from the KV cache — no FP32 staging buffer — saving ~14 MB per call and one memory pass.

### Performance (Qwen3-0.6B, Q4_0-FP32, 8 threads)

**Long prefill (1003 tokens) + 32-token decode:**

| Device | Flash OFF (prefill ms / TPS) | Flash ON (prefill ms / TPS) | Prefill speedup |
|--------|------------------------------|-----------------------------|-----------------|
| S25 Ultra | 2967 ms / 338 TPS | 1491 ms / 672 TPS | **2.0×** |
| S26 Ultra | 3047 ms / 329 TPS | 1768 ms / 567 TPS | **1.7×** |
| x86 Ryzen 7 | 5994 ms / 167 TPS | 6242 ms / 160 TPS | ~equal |

**Short prefill (36 tokens), x86 Q4:**

| | Prefill | Generation |
|--|---------|------------|
| Flash OFF | 1551 ms (23 TPS) | 10620 ms |
| Flash ON  | 128 ms (281 TPS) | 441 ms |
| Speedup | **12×** | **24×** |

**V-JEPA ViT-B 16-frame (N=4608), x86 FP32:**
- Reference: 27.0 s → Flash: **13.9 s (~2× faster)**

---

### Kernels added (NEON, `ENABLE_FP16` only)

- `nntr_gemm_q4_0_4x8_q8_0_fp16` / `nntr_gemv_q4_0_4x8_q8_0_fp16`: Q4_0 weight × FP16 activation → FP16 output (`nntr_ggml_impl_neon.cpp`)
- Q6_K dispatch: `HalfTensor::dotQnK` routes Q6_K weights through `nntr_vec_dot_q6_K_q8_K` with FP16 quantised activation

### Layer fixes enabling the full FP16 activation path

- **`InputLayer`**: removed the force-promote of FP32 graph input to the model activation dtype; removed the non-FP32 in-place guard
- **`EmbeddingLayer` / `TokenTypeEmbeddingLayer`**: removed the `finalize()` check that rejected non-FP32 input dimensions
- **`RMSNorm` / `ReshapedRMSNorm`**: restored the FP16 branch calling `rms_norm_wrt_width_fp16_intrinsic`; gamma cast to activation dtype at the multiply site
- **`FCLayer` bias**: bias stays FP32, cast to activation dtype at the add site
- **`HalfTensor::dot` Q4_0**: was an unreachable `break`; now routes to `HalfTensor::dotQnK` → `gemm_q4_0_fp16`

### `tie_word_embedding` lm_head fixes

- Corrected Q6_K logits: explicit per-row dequant + dot with FP32 upcast; logits output forced to FP32 regardless of activation dtype
- Restored Q6_K decode throughput: reverts workaround that ran Q6_K through the slow per-row path; routes it back through the fused GEMV

  | Device | Before | After |
  |--------|--------|-------|
  | S26 Ultra (FP16) | 38 tok/s | **79 tok/s** |
  | S26 Ultra (FP32) | 33 tok/s | **69 tok/s** |

---

## Tests

- New `Qwen3Fp16` test suite in `unittest_causallm_qwen3.cpp` (gated `ENABLE_FP16`): Q4_0 × FP16 activation forward pass, tolerance 2e-3
- `makeTinyQ40Fp16DataType()` helper + `toTensorDataType("FP16")` utility in `causallm_test_utils`
- `IterationForGolden` fix: correctly reads FP16-input golden files

---

## Key files changed

| Area | Files |
|------|-------|
| Flash attention | `mha_core.cpp/.h`, `cpu_backend_gemm_decl.h` (new), `FLASH_ATTENTION.md` (new) |
| ARM GEMM kernels | `cpu_backend/arm/hgemm/hgemm.cpp/.h` (new), `nntr_ggml_impl_neon.cpp` |
| x86 GEMM kernels | `cpu_backend/x86/avx2_impl.cpp/.h` |
| FP16 ggml dispatch | `ggml_interface_fp16.cpp`, `nntr_ggml_impl.h` |
| Layer fixes | `input_layer.cpp`, `embedding.cpp`, `half_tensor.cpp` |
| lm_head | `tie_word_embedding.cpp` |
| Config/model | `transformer.cpp/.h`, `qwen3_causallm.cpp` |

### CODE REVIEW

File: Applications/CausalLM/layers/mha_core.cpp [hunk 1]
```diff
@@ -221,6 +224,14 @@ void MHACoreLayer::finalize(nntrainer::InitLayerContext &context) {
 
   /** Is Causal */
   is_causal = std::get<props::IsCausal>(mha_core_props).get();
+  use_gemm_attention = std::get<props::UseGemmAttention>(mha_core_props).get();
+#if !(ENABLE_FP16 && defined(__ANDROID__))
+  // The GEMM / flash-attention path (gemm_attention) is built and verified
+  // only for the ARM FP16 device build. On x86 / non-FP16 builds the AVX2
+  // flash path is unvalidated and NaNs on the wide V-JEPA encoder logits, so
+  // fall back to the reference attention kernels there.
+  use_gemm_attention = false;
```
Comment (baek2sm) [inline Applications/CausalLM/layers/mha_core.cpp:232]: On x86, use_gemm_attention is always forced to false. Were the x86 "Flash ON" results in the PR description actually measured with flash attention, or are they from before this gate was added?

File: Applications/CausalLM/layers/mha_core.cpp [hunk 2]
```diff
@@ -781,6 +802,365 @@ void MHACoreLayer::one_batch_incremental_forwarding(
                                 cache_to);
 }
 
+void MHACoreLayer::gemm_attention(nntrainer::Tensor &query_step,
```
Comment (jaemini-shin) [inline Applications/CausalLM/layers/mha_core.cpp:805]: Is it possible to delegate this method to AVX2 and neon kernel?

**Comment by jaemini-shin** [comment]
Is it possible to separate this PR into two PRs?
* Add flash attention
* Add FP16 activation

Since nntrainer attention has used FP16 QKVO, adding FP16 activation doesn't affect to flash attention. So if it's possible to make two different PRs, it'd be better to review and check the performance.

File: nntrainer/tensor/cpu_backend/arm/hgemm/hgemm.cpp [hunk 3]
```diff
@@ -195,3 +195,136 @@ void hgemm_K1(const __fp16 *A, const __fp16 *B, __fp16 *C, unsigned int M,
...
+// the rows are contiguous).
+//
+// Requires FEAT_FHM (asimdfhm in /proc/cpuinfo). The target attribute pulls
+// the fp16fml ISA extension in only for this function so the rest of the TU
+// can stay on the build-wide -march flags.
+// Single-output FP16xFP16->FP32 dot, used for the M/N tails of the blocked
+// kernel below. The 4x2 block reproduces this exact accumulation order per
+// output, so blocked and tail results are bit-identical (no token drift).
+__attribute__((
+  target("arch=armv8.2-a+fp16+fp16fml+dotprod+i8mm"))) static inline float
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/arm/hgemm/hgemm.cpp:220]:
```suggestion
static inline float
```
With this attribute, it will not work for armv9.2-a config.

File: nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_fp16.cpp [hunk 4]
```diff
@@ -352,10 +351,13 @@ void __ggml_gemm_q6_K(const unsigned int M, const unsigned int N,
 
       const void *const B_data = (void *)((char *)B + B_row_data_offset);
 
-      nntr_vec_dot_q6_K_q8_K(K, &C32_ptr[thread_job], bs, B_data, bx,
-                             quantized_A_data, by, nrc);
+      float result;
+      nntr_vec_dot_q6_K_q8_K(K, &result, bs, B_data, bx, quantized_A_data, by,
+                             nrc);
+      C[thread_job] = (_FP16)result;
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface_fp16.cpp:357]: This line might result consecutive false sharing, we should compare the result for this change.

File: nntrainer/tensor/cpu_backend/arm/arm_compute_backend.h [hunk 5]
```diff
@@ -35,6 +35,30 @@
...
+ *       targets -march=...+fp16), since __fp16 is not a usable type on ARM
+ *       targets built without that extension (e.g. Tizen armv7l/aarch64,
+ *       which do not support fp16 NEON).
+ */
+void shgemm(const unsigned int TStorageOrder, bool TransA, bool TransB,
+            const unsigned int M, const unsigned int N, const unsigned int K,
+            const float alpha, const float *A, const unsigned int lda,
+            const __fp16 *B, const unsigned int ldb, const float beta, float *C,
+            const unsigned int ldc);
+#endif
```
Comment (jaemini-shin) [inline nntrainer/tensor/cpu_backend/arm/arm_compute_backend.h:60]: Is this declaration necessary? AFAIK, `armv7l` builds `libnntrainer.so` with `enable-fp16=false`.

File: Applications/CausalLM/layers/mha_core.cpp [hunk 6]
```diff
@@ -781,6 +802,365 @@ void MHACoreLayer::one_batch_incremental_forwarding(
...
+        //     No FP32 copy of Q is materialized.
+        //   - !q_fp16: FP32 Q × FP16 K via shgemm / avx2 / sgemm.
+        // The softmax below reads S (FP32) and stores normalized FP16 probs to
+        // Sp16 for the AV hgemm_f16xf16_f16.
+#if defined(__x86_64__) || defined(__i386__)
+        nntrainer::avx2::hsgemm_fp16bits_avx2(bq, bk, d, inv_sqrt, Qp_fp32, d,
+                                              Kp, d, /*TransB=*/true, S.data(),
+                                              bk);
+#elif defined(__ARM_NEON)
+        if (q_fp16) {
```
Comment (jaemini-shin) [inline Applications/CausalLM/layers/mha_core.cpp:1002]: To avoid compile issue, it'd be better to branch kernels in compile time such as `ENABLE_FP16` not a boolean variable.

**Review by jaemini-shin — COMMENTED** [review]
I left some comments, PTAL!

**Comment by baek2sm** [comment]
This PR is seperated to PR #4044 and #4045 , so I'll close it
